File: java/src/main/java/io/binac/leetcode/SwapNodesInPairs.java
package io.binac.leetcode;

import io.binac.leetcode.util.ListNode;

/**
 * Given a linked list, swap every two adjacent nodes and return its head.
 * <p>
 * <p>Example:
 * <blockquote><pre>
 *     Given 1->2->3->4, you should return the list as 2->1->4->3.
 * </blockquote></pre>
 * Note:
 * <p>
 * <p>Your algorithm should use only constant extra space.
 * <p>You may not modify the values in the list's nodes, only nodes itself may be changed.
 */
public class SwapNodesInPairs {
    public static class Solution1 {
        public ListNode swapPairs(ListNode head) {
            ListNode dummy = new ListNode(0);
            dummy.next = head;
            ListNode pre = dummy, p = head;
            while (p != null && p.next != null) {
                pre.next = p.next;
                p.next = p.next.next;
                pre.next.next = p;
                pre = p;
                p = p.next;
            }
            return dummy.next;
        }
    }
}

File: python3/leetcodepy/swap_nodes_in_pairs.py
"""
Given a linked list, swap every two adjacent nodes and return its head.



Example 1:

https://assets.leetcode.com/uploads/2020/10/03/swap_ex1.jpg

Input: head = [1,2,3,4]
Output: [2,1,4,3]

Example 2:

Input: head = []
Output: []

Example 3:

Input: head = [1]
Output: [1]


Constraints:

The number of nodes in the list is in the range [0, 100].
0 <= Node.val <= 100


Follow up: Can you solve the problem without modifying the values in the list's nodes? (i.e., Only nodes themselves may be changed.)
"""

from .utils import ListNode


class Solution1:
    def swapPairs(self, head: ListNode) -> ListNode:
        dummy = ListNode(0, head)
        tail, p1, p2 = dummy, head, None
        while p1 and p1.next:
            p2 = p1.next
            p1.next = p2.next
            p2.next = p1
            tail.next = p2
            tail = p1
            p1 = p1.next
        return dummy.next
